### ISSUE TITLE
jax.experimental.sparse.BCOO sum over axis does not reduce number of data entries

### ISSUE THREAD

**Issue by GarfieldGa** [jax-ml/jax#26559]
### Description

I was using the experimental feature BCOO for a project, and I found that the sum over axis function of BCOO matrices are not working as expected. For example, if I have a matrix of shape (50,50). By broadcasting it to the shape of (10,50,50) using bcoo_broadcast_in_dim, and then summing over the batched axis should leave the number of indices and non-zero data entries unchanged. However, when I do the testing on my local machine, this is not the case. The issue can be recreated using the following code:
```
import jax
import jax.numpy as jnp
from jax.experimental.sparse import BCOO, bcoo_reduce_sum, bcoo_broadcast_in_dim

if __name__ == '__main__':
    A = jnp.zeros((50,50))
    for i in range(50):
        A = A.at[i,i].set(1.0)
        if i > 0:
            A = A.at[i,i-1].set(-0.5)
        if i < 49:
            A = A.at[i,i+1].set(-0.5)
    # Create a BCOO matrix from a dense matrix (50,50)
    A = BCOO.fromdense(A)
    print('A shape',A.shape, 'indices shape', A.indices.shape, 'indicies max', A.indices.max(), 'data shape', A.data.shape)
    # Broadcast the BCOO matrix to shape (10,50,50)
    resulting_shape = (10,50,50)
    A = bcoo_broadcast_in_dim(A, shape=resulting_shape, broadcast_dimensions=[1,2])
    print('A shape',A.shape, 'indices shape', A.indices.shape, 'indicies max', A.indices.max(), 'data shape', A.data.shape)
    # Sum over the batched axis to shape (50,50)    
    A = A.sum(axis=0)
    # A = bcoo_reduce_sum(A, axes=[0])
    print('A shape',A.shape, 'indices shape', A.indices.shape, 'indicies max', A.indices.max(), 'data shape', A.data.shape)
```
The printout looks like below:

![Image](https://github.com/user-attachments/assets/e7a01123-323c-48ae-9821-cda2649894f6)
We can see that the number of indices and data entries are 10 times larger.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.38
jaxlib: 0.4.38
numpy:  1.26.2
python: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Windows', node='MyDesktop', release='10', version='10.0.19045', machine='AMD64')

$ nvidia-smi
Sun Feb 16 13:31:59 2025
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 538.92                 Driver Version: 538.92       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA RTX 3500 Ada Gene...  WDDM  | 00000000:01:00.0 Off |                  Off |
| N/A   55C    P4              17W / 115W |   2132MiB / 12282MiB |     19%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

**Comment by jakevdp** [comment]
Hi - thanks for the report. This is working as expected. JAX sparse representations require the number of indicies to be statically known. This means that in the general case, the number of indices to store the result of a sparse reduction is unknowable. For that reason, `jax.experimental.sparse` implements `sum` without doing any explicit reduction in indices. If you wish to trigger such a reduction explicitly, you can call `bcoo.sum_duplicates()`

**Comment by GarfieldGa** [comment]
Thanks for the explanation, this is really helpful!
